cumulus/pallets/collator-selection/src/lib.rs

```diff
@@ -554,6 +596,168 @@ pub mod pallet {
 			Self::deposit_event(Event::InvulnerableRemoved { account_id: who });
 			Ok(())
 		}
+
+		/// Update the candidacy bond of collator candidate `origin` to a new amount `new_deposit`.
+		///
+		/// Setting a `new_deposit` that is lower than the current deposit while `origin` is
+		/// occupying a top-`DesiredCandidates` slot is not allowed.
+		///
+		/// This call will fail if `origin` is not a collator candidate, the updated bond is lower
+		/// than the minimum candidacy bond, and/or the amount cannot be reserved.
+		#[pallet::call_index(7)]
+		#[pallet::weight(T::WeightInfo::update_bond(T::MaxCandidates::get()))]
+		pub fn update_bond(
+			origin: OriginFor<T>,
+			new_deposit: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+			ensure!(new_deposit >= <CandidacyBond<T>>::get(), Error::<T>::DepositTooLow);
+			// The function below will try to mutate the `CandidateList` entry for the caller to
+			// update their deposit to the new value of `new_deposit`. The return value is the
+			// position of the entry in the list, used for weight calculation.
+			let length =
+				<CandidateList<T>>::try_mutate(|candidates| -> Result<usize, DispatchError> {
+					let mut idx = candidates
+						.iter()
+						.position(|candidate_info| candidate_info.who == who)
+						.ok_or_else(|| Error::<T>::NotCandidate)?;
+					let candidate_count = candidates.len();
+					let old_deposit = candidates[idx].deposit;
+					if new_deposit > old_deposit {
+						T::Currency::reserve(&who, new_deposit - old_deposit)?;
+					} else if new_deposit < old_deposit {
+						// Casting `u32` to `usize` should be safe on all machines running this.
+						ensure!(
+							idx.saturating_add(<DesiredCandidates<T>>::get() as usize) <
+								candidate_count,
+							Error::<T>::InvalidUnreserve
+						);
+						T::Currency::unreserve(&who, old_deposit - new_deposit);
+					} else {
+						return Err(Error::<T>::IdenticalDeposit.into())
+					}
+					candidates[idx].deposit = new_deposit;
+					// Since `idx` is at most `T::MaxCandidates - 1`, because it's the position of
+					// an entry in a list that is at most `T::MaxCandidates` long, it will never be
+					// an out of bounds index access. Furthermore, the increment will not be an
+					// overflow since `idx` is `usize` and it can be at most `T::MaxCandidates`,
+					// which is `u32`.
+					if new_deposit > old_deposit && idx < candidate_count {
+						// Since the lowest value for `idx` would be `0`, the first element in the
+						// list, and the fact we do an increment of `idx` before iterating, we
+						// ensure the `idx - 1` list access in the loop is not an underflow.
+						idx += 1;
+						while idx < candidate_count && candidates[idx].deposit < new_deposit {
+							candidates.as_mut().swap(idx - 1, idx);
```

Comment:
you could do this in one iteration, calculate new_deposit, iterate over all items (which sorted), look for two position, new_pos and old_pos, change position. 
with idx inc and dec this look complex, and amount of comments proof it.

Reply:
> you could do this in one iteration, calculate new_deposit, iterate over all items (which sorted), look for two position, new_pos and old_pos, change position.

Not that it matters too much since we agreed [here](https://github.com/paritytech/cumulus/pull/2960#discussion_r1303216042) that performance isn't really relevant with 20 entries in a `vec`, but that would require a `remove` and an `insert`, which I don't really like considering the alternative below.

> with idx inc and dec this look complex, and amount of comments proof it.

The comments are there to explain why the code is safe. They are not there to explain how it works. Also, it's just an ascending or descending pivot function, I really don't think it's that complex. I'd be willing to take that and extract it into a pivot function if you think it would look better.

Reply:
I think it is better without additional abstractions

Reply:
Refactored to get `new_pos` and `old_pos` in one pass.